v0/customerio/transform.js

```diff
@@ -0,0 +1,222 @@
+const _ = require("lodash");
+const get = require("get-value");
+const set = require("set-value");
+const btoa = require("btoa");
+
+const { EventType, SpecedTraits, TraitsMapping } = require("../../constants");
+const {
+  removeUndefinedValues,
+  defaultPostRequestConfig,
+  defaultPutRequestConfig,
+  defaultRequestConfig,
+  isPrimitive
+} = require("../util");
+
+const {
+  IDENTITY_ENDPOINT,
+  USER_EVENT_ENDPOINT,
+  ANON_EVENT_ENDPOINT,
+  DEVICE_REGISTER_ENDPOINT,
+  DEVICE_DELETE_ENDPOINT
+} = require("./config");
+
+const deviceRelatedEventNames = [
+  "Application Installed",
+  "Application Opened",
+  "Application Uninstalled"
+];
+const deviceDeleteRelatedEventName = "Application Uninstalled";
+
+// Get the spec'd traits, for now only address needs treatment as 2 layers.
+// populate the list of spec'd traits in constants.js
+const populateSpecedTraits = (payload, message) => {
+  //console.log(message);
+  SpecedTraits.forEach(trait => {
+    const mapping = TraitsMapping[trait];
+    const keys = Object.keys(mapping);
+    keys.forEach(key => {
+      set(payload, key, get(message, mapping[key]));
+    });
+  });
+};
+
+function responseBuilder(message, evType, evName, destination) {
+  const rawPayload = {};
+  let endpoint;
+  let requestConfig = defaultPostRequestConfig;
+  //console.log(message);
+  let userId =
+    message.userId && message.userId != "" ? message.userId : undefined;
+
+  const response = defaultRequestConfig();
+  response.userId = message.userId ? message.userId : message.anonymousId;
+
+  if (evType === EventType.IDENTIFY) {
+    // populate speced traits
+    populateSpecedTraits(rawPayload, message);
+    if (message.context.traits) {
+      const traits = Object.keys(message.context.traits);
+      traits.forEach(trait => {
+        //populate keys other than speced traits
+        if (!SpecedTraits.includes(trait)) {
+          set(rawPayload, trait, get(message, "context.traits." + trait));
+        }
+      });
+    }
+
+    if (message.user_properties) {
+      const userProps = Object.keys(message.user_properties);
+      userProps.forEach(prop => {
+        let val = get(message, "user_properties." + prop);
+        // send only top level keys
+        if (isPrimitive(val)) {
+          set(rawPayload, prop, val);
+        }
+      });
+    }
+
+    set(
+      rawPayload,
+      "created_at",
+      Math.floor(new Date(message.originalTimestamp).getTime() / 1000)
+    );
+    //console.log(rawPayload);
+
+    if (userId) {
```

Comment:
Let's move this check to the beginning of the if block. As a practice, whenever we can return from a function those checks should be at the top.

Reply:
Done